File: RIONet/ga2230net/src/main/java/org/ga2230net/MappedListenerThread.java
package org.ga2230net;

import java.io.IOException;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Map;
import java.util.concurrent.locks.ReentrantLock;
import java.util.LinkedHashMap;

/**
 * A thread that listens for incoming packets from a sender and adds them to a queue
 */
public class MappedListenerThread extends Thread {
    private final ReentrantLock lock;
    private final Map<String, Queue<Packet>> packetTable;
    private final ListenerSocket listenerSocket;
    private boolean running;
    private static MappedListenerThread listenerThread;

    /**
     * create a new listener thread
     * @param port the port to listen on
     * @param builder the packet builder to use to build packets
     * @throws IOException if an error occurs while creating the listener sockets
     */
    private MappedListenerThread(int port, PacketBuilder builder) throws IOException {
        lock = new ReentrantLock();
        packetTable = new LinkedHashMap<>();
        listenerSocket = new ListenerSocket(port, builder);
    }

    /**
     * initializes the single ListenerThread instance
     * @param port the port to listen on
     * @param builder the builder to use for parsing raw packets
     * @throws IOException if an IO error occurred
     */
    public static void initListener(int port, PacketBuilder builder) throws IOException {
        listenerThread = new MappedListenerThread(port, builder);
    }


    /**
     * @return the single ListenerThread instance, null if the listener wasn't initialized
     */
    public static MappedListenerThread getInstance() {
        return listenerThread;
    }

    public void run() {
        running = true;
        while (running) {
            try {
                Packet packet = listenerSocket.getPacket();
                addPacket(packet);
            } catch (IOException e) {
                e.printStackTrace();
                running = false;
            }
        }
    }

    /**
     * get the next n packets from the received packets queue of the given header
     *
     * @return a Packet array from the header's packet queue, returns null if
     *         the queue is empty
     */
    public Packet[] getPackets(String header, int numOfPackets) {
        lock.lock();
        try {
            Queue<Packet> pq = packetTable.get(header);
            Packet[] packets = new Packet[numOfPackets];
            for(int i = 0; i < numOfPackets; i++){
                packets[i] = pq.poll();
            }
            return packets;
        } finally {
            lock.unlock();
        }
    }

    /**
     * @return an array of all packets left in the packet queue of the given header
     */
    public Packet[] getPackets(String header) {
        Packet[] packets = null;
        lock.lock();
        try {
            Queue<Packet> pq = packetTable.get(header);
            if (pq != null) {
                packets = new Packet[pq.size()];
                for(int i = 0; i < packets.length; i++){
                    packets[i] = pq.poll();
                }
            }
        } finally {
            lock.unlock();
        }
        return packets;
    }

    /**
     * flushes the packet queue of the given header and thus deleting all packets in it
     */
    public void flushPacketsQueue(String header) {
        lock.lock();
        try {
            Queue<Packet> pq = packetTable.get(header);
            pq.clear();
        } finally {
            lock.unlock();
        }
    }

    public boolean isRunning() {
        return running;
    }

    private void addPacket(Packet packet) {
        lock.lock();
        try {
            Queue<Packet> pq = packetTable.get(packet.getHeader());
            if (pq == null) {
                pq = new LinkedList<>();
                pq.add(packet);
                packetTable.put(packet.getHeader(), pq);
            } else {
                if (packet.isSingleInstance()) {
                    for (Packet p : pq) {
                        if (p.getHeader().equals(packet.getHeader())) {
                            pq.remove(p);
                            break;
                        }
                    }
                }
                pq.add(packet);
            }
        } finally {
            lock.unlock();
        }
    }
}

File: PyNet/ga2230Net/MappedListenerThread.py
from __future__ import annotations

from threading import Thread, Lock
from queue import Queue, Empty
import socket

from typing import List, Dict

from .ListenerSocket import ListenerSocket
from .Packet import Packet
from .PacketBuilder import PacketBuilder


class MappedListenerThread(Thread):
    """A thread handler for a listener that will continuously listen
      on a specified port and insert all data received into a
      map of headers to packet queues
    """

    __instance: MappedListenerThread = None

    def __new__(cls, port: int, packet_builder: PacketBuilder) -> None:
        """
        :param port: the port to listen on
        :type port: int
        :param packet_builder: the packet builder to use
        :type packet_builder: PacketBuilder
        """
        if cls.__instance is None:
            cls.__instance = super(MappedListenerThread, cls).__new__(cls)
            cls.__instance.listener_socket: ListenerSocket = ListenerSocket(
                port, packet_builder
            )
            cls.__instance.packet_map: Dict[Queue] = {}
            cls.__instance.builder: PacketBuilder = packet_builder
            cls.__instance.running = True
            cls.__instance.mutex = Lock()
        return cls.__instance

    def __init__(self, *args, **kwargs) -> None:
        super().__init__()

    def get_instance(cls) -> MappedListenerThread:
        """Returns the instance of the listener thread

        :return: the instance of the listener thread
        :rtype: MappedListenerThread
        """
        return cls.__instance

    def run(self) -> None:
        while self.running:
            try:
                new_packet: Packet = self.listener_socket.get_packet()
                self._add_to_queue(new_packet)
            except Exception as e:
                print(e)
                self.running = False

    def get_packets(self, header: str,
                    num_of_packets: int = None) -> List[Packet]:
        """Retrieves a list of packets from the packet map according
          to the given header.
        Returns all packets if a number wasn't specified.

        :param header: the header of the packets
        :type header: str
        :param num_of_packets: the number of packets to retrieve
        :type num_of_packets: int
        :return: a list of packets of a specified header
        :rtype: List[Packet]
        :raises ValueError: if the header is not found in the packet schemes
        """

        if header not in self.builder.packet_schemes.keys():
            raise ValueError(f"Header {header} not found in packet schemes.")

        if num_of_packets is None:
            num_of_packets = self.packet_map[header].qsize()

        if header not in self.packet_map.keys():
            return []

        packets = []
        with self.mutex:
            for _ in range(num_of_packets):
                try:
                    packets.append(self.packet_map[header].get(block=False))
                except Empty:
                    break

        return packets

    def get_packet(self, header: str) -> Packet:
        """Retrieves a single packet from the packet queue of the
          specified header

        :return: a packet
        :rtype: Packet
        """
        return next(iter(self.get_packets(header, 1)), None)

    def flush_packets(self, header: str = None) -> None:
        """Flushes the packet queue of the specified packet header
        if a header wasn't specified, it will flush all packet queues in the
          map
        """

        if header not in self.packet_map.keys():
            return

        with self.mutex:
            if header is None:
                for q in self.packet_map.values():
                    q.queue.clear()
            else:
                self.packet_map[header].queue.clear()

    def is_running(self) -> bool:
        """Returns whether the thread is running

        :return: whether the thread is running
        :rtype: bool
        """
        return self.running

    def _add_to_queue(self, packet: Packet) -> None:
        """Adds a packet to the packet queue

        :param packet: the packet to add
        :type packet: Packet
        """
        if packet.header not in self.packet_map.keys():
            self.packet_map[packet.header] = Queue()
        with self.mutex:
            if packet.is_single_instance():
                self.packet_map[packet.header].get()
            self.packet_map[packet.header].put(packet)
